File: rust/src/bin/sum_of_even_squares.rs
fn sum_of_even_squares(a: &[i32]) -> i32 {
    a.iter()
        .filter(|&&x| x % 2 == 0)
        .map(|&x| x * x)
        .sum()
}

fn main() {
    let a = vec![1, 2, 3, 4, 5, 6, 7, 8];
    assert_eq!(sum_of_even_squares(&a), 120);
    assert_eq!(sum_of_even_squares(&a[1..=5]), 56);
}

File: haskell/sum_of_even_squares.hs
import Control.Monad (guard)

sumOfEvenSquares :: [Int] -> Int
sumOfEvenSquares = sum . map (^ 2) . filter even

main = do 
    guard $ sumOfEvenSquares [] == 0
    guard $ sumOfEvenSquares [1] == 0
    guard $ sumOfEvenSquares [8] == 64
    guard $ sumOfEvenSquares [1, 2, 3, -4] == 20
    guard $ sumOfEvenSquares [1, 3, 5, 7] == 0
    guard $ sumOfEvenSquares [2, 4, 6, 8] == 120

